### ISSUE TITLE
Make it possible to configure rule-graph from shakeOptions

### ISSUE THREAD

**Issue by nkaretnikov** [ndmitchell/shake#467]
My graph has three rows and lots of nodes are located next to each other, which makes it impossible to follow the arrows between them. I've tried playing with what appears to be the relevant [code](https://github.com/ndmitchell/shake/blob/e4f522b92013b59c7367593b195232f3e67f5889/html/shake.js#L776), but I'm having trouble understanding it. Moreover, the process is very tedious: I need to recompile shake (~100 modules), my application, and finally build the report. (I didn't succeed after spending ~20 minutes on it.) Would be really nice to be able to do this via `shakeOptions` or a similar mechanism. How about moving the stuff under `html` to GHCJS?

Perhaps the graph is not the most important feature, but it's quite useful, and I'm looking forward to using it to spot bottlenecks in my system (as advertised in the whitepaper). I think that moving everything to Haskell will pay off in the long term. It's like with Bash scripts: they tend to grow, then become a complete mess, so it's better to use Haskell from the start :).

**Comment by ndmitchell** [comment]
I agree that using Haskell is better than Javascript (or bash etc). But, for user interfaces, HTML is the only reasonable solution without a massive load of dependencies. Once you are stuck to HTML, Javascript is the only sensible answer - and for that I translate from TypeScript (GHCJS is both too hard to configure and produces bloated and slow output compared to TypeScript).

Fortunately, you don't need to recompile Shake to get different graph mappings! Given an HTML output report, tweaking the textbox at the top and pressing Run will give you a different graph. As an example, see:

* [Complete graph](https://cdn.rawgit.com/ndmitchell/shake/35fbe03c8d3bafeae17b58af89497ff3fdd54b22/html/report.html?mode=rule-graph)
* [By extension](https://cdn.rawgit.com/ndmitchell/shake/35fbe03c8d3bafeae17b58af89497ff3fdd54b22/html/report.html?mode=rule-graph&query=name%28/%28%5C.%5Ba-zA-Z0-9_%5D%2B%29%24/%2C%20%22%3F%3F%3F%22%29)
* [By directory](https://cdn.rawgit.com/ndmitchell/shake/35fbe03c8d3bafeae17b58af89497ff3fdd54b22/html/report.html?mode=rule-graph&query=name%28/%5E%28%5Ba-z%5D%2B%29%5C//%2C%20%22%3Croot%3E%22%29)
* [Omitting intermediate structure](https://cdn.rawgit.com/ndmitchell/shake/35fbe03c8d3bafeae17b58af89497ff3fdd54b22/html/report.html?mode=rule-graph&query=name%28%22.cc%22%29%20%7C%7C%20name%28%22.exe%22%29)

If you post your profile report from an unmodified Shake I should be able to show you how to render it in some more meaningful way.

**Comment by nkaretnikov** [comment]
@ndmitchell Thanks for the prompt reply! Glad that it's possible to tweak, but I still want to get a readable graph without filtering. Mine is similar to the "complete graph" shown above. For example, I'd like to move the rows away from each other, is it possible without recompiling?

**Comment by ndmitchell** [comment]
It isn't possible even _with_ recompiling - that's just the GraphViz output. The hope is that #353 and switching the graph library will make it possible, and also make the profiles smaller.

**Comment by nkaretnikov** [comment]
Also, I configure the browser to use my own fonts instead of the provided ones, which doesn't play nicely with the default graph settings (node-bubbles are too small for their labels). I can live with this, but it's just something to consider, too.

Regarding GraphViz: if it's an ordinary GraphViz and not a similarly named package (I couldn't find where it's called in the code), there are options to change the layout of nodes (I used that in the past). I have a script at work that looks quite nice, I look up the options I used tomorrow and report back.

**Comment by ndmitchell** [comment]
It is ordinary GraphViz, and if there are any small tweaks you know of to the rendered output that can be applied in the dot output you should be able to make them directly to the profile.html, saving you the recompile cycle. I construct the dot script [here](https://github.com/ndmitchell/shake/blob/master/html/shake.js#L776), and render it [here](https://github.com/ndmitchell/shake/blob/master/html/shake.js#L795) - but you'll find the contents of `shake.js` inlined into your profile output.

My guess is GraphViz isn't going to place nicely with your fonts since it's a really nasty LLVM compiled to a static byte array trick, so really not browser native. In contrast, D3 should work properly.

**Comment by nkaretnikov** [comment]
> It is ordinary GraphViz, and if there are any small tweaks you know of to the rendered output that can be applied in the dot output you should be able to make them directly to the profile.html, saving you the recompile cycle.

Ah, right! Thanks for the hint. The options I was thinking of are `dpi=1500,splines=\"true\",overlap=\"scalexy\"`, but that was for the sfdp command, so it might work differently. I tried passing that to the graph attribute in the report file, but then I got a blank page instead of the graph (probably it was too large or something). As a workaround I just increased `ranksep` a bit. I do wonder, however, why `overlap` doesn't work. Maybe it's the `dpi` option that actually matters, hm...
